### PR TITLE
[3040] Migrate UI comp `ViewModels` to `OfflinePlugin`

### PR DESCRIPTION
**Pull request by MarinTolic**
### 🎯 Goal

Migrate the UI Components building blocks that are dependant on `ChatDomain`  to `OfflinePlugin`.

### 🛠 Implementation details

- Migrated UI Components ViewModels
- Migrated UI Components Views
- Migrated `stream-chat-android-ui-components` and `stream-chat-android-ui-common` extension functions etc. 
- Migrated the UI Components sample module

### 🎨 ~~UI Changes~~

**No UI changes**

### 🧪 Testing

Unfortunately, the truth is that this PR requires a complete test of most if not all of our features.
However, this is a list of things that need special attention.

**MessageListHeaderView:**

- Switches description when switching from normal to thread mode
- Online indicator works properly
- Avatar updates correctly
- User online status updates correctly
- List of members updates correctly

**MessageInputViewModel:** 

- Posting a message works
- Correct commands are displayed
- The cooldown interval is correct, where applicable
- Maximum message length is respected
- Replies work as expected
- Posting to threads works as expected
- Editing messages works as expected
- Mentions work as expected
- Is in the correct `ChatMode`

**ChannelActionsViewModel:**

- Displays the correct channel information such as name, number of members and their names etc
- Performs the available actions such as deleting a channel or leaving a group correctly

**ChannelListHeaderViewModel:**

- Current user is properly displayed
- Connection state is properly displayed

**TypingIndicatorViewModel**

- Start typing a message and make sure that the indicator is visible no another device.

**MessageListViewModel:**

- Is in the correct `ChatMode`
- Messages are properly marked as read
- The user is correct
- The channel an it's state are correct

**ChannelListViewMode:**

- Channel mutes are correctly displayed
- List loads correctly
- List requests new data correctly

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] ~~New code is covered by unit tests~~ (This effort will be moved to a separate PR)
- [ ] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

delete

### CODE REVIEW

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt [hunk 1]
```diff
@@ -1,70 +1,95 @@
...
-                _typingUsers.addSource(channelController.typing) { typingEvent ->
-                    _typingUsers.value = typingEvent.users
-                }
-            } else {
-                logger.logE("Could not watch channel with cid: $cid. Error: ${channelControllerResult.error()}")
-            }
+    /**
+     * A list of users who are currently typing.
+     */
+    public val typingUsers: LiveData<List<User>> =
```
Comment (adasiewiczr) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt:54]: Should we expose this `LiveData` if it can be already accessed through `channelState`? 🤔

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt [hunk 2]
```diff
@@ -1,70 +1,95 @@
...
+        Transformations.map(channelState.typing.asLiveData()) { typingEvent ->
+            typingEvent.users
         }
-    }
 
+    /**
+     * A list of [Channel] members.
+     */
+    public val members: LiveData<List<Member>> = channelState.members.asLiveData()
```
Comment (adasiewiczr) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt:62]: Same here

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt [hunk 3]
```diff
@@ -18,83 +20,121 @@ import io.getstream.chat.android.client.models.Message
...
 public class MessageInputViewModel @JvmOverloads constructor(
     private val cid: String,
-    private val chatDomain: ChatDomain = ChatDomain.instance(),
     private val chatClient: ChatClient = ChatClient.instance(),
 ) : ViewModel() {
+
+    /**
+     * Holds information about the current channel and is actively updated.
+     */
+    public val channelState: ChannelState =
```
Comment (adasiewiczr) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt:45]: Should we make it public? If yes, should we also expose other properties that can be accessed through `channelState`?

File: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/add/AddChannelViewModel.kt [hunk 4]
```diff
@@ -112,7 +114,7 @@ class AddChannelViewModel : ViewModel() {
             return
         }
         viewModelScope.launch(Dispatchers.IO) {
-            val currentUserId = chatDomain.user.value?.id ?: error("User must be set before create new channel!")
+            val currentUserId = chatClient.globalState.user.value?.id ?: error("User must be set before create new channel!")
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/add/AddChannelViewModel.kt:117]: We can also call `chatClient.getCurrentUser()?.id`

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/typing/viewmodel/TypingIndicatorViewModel.kt [hunk 5]
```diff
@@ -1,30 +1,38 @@
...
     cid: String,
-    chatDomain: ChatDomain = ChatDomain.instance(),
+    chatClient: ChatClient = ChatClient.instance(),
 ) : ViewModel() {
 
-    private val _typingUsers = MediatorLiveData<List<User>>()
+    /**
+     * Holds information about the current channel and is actively updated.
+     */
+    public val channelState: ChannelState =
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/typing/viewmodel/TypingIndicatorViewModel.kt:28]: Should we make it public? 🤔

File: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/add/AddChannelViewModel.kt (hunk 4 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/add/AddChannelViewModel.kt:117]: I would prefer to use globalState directly to request to user and added GlobalState as a dependency of this ViewModel, then we avoid accessing globalState thought ChatClient.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt [hunk 6]
```diff
@@ -18,83 +20,121 @@ import io.getstream.chat.android.client.models.Message
...
+    public val cooldownInterval: LiveData<Int> =
+        Transformations.map(channelState.channelData.asLiveData()) { channelData ->
+            channelData.cooldown
+        }
+
+    /**
+     * The maximum length of a message that can be typed in the message input.
+     */
+    public val maxMessageLength: LiveData<Int> =
+        Transformations.map(channelState.channelConfig.asLiveData()) { config ->
```
Comment (leandroBorgesFerreira) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt:73]: What about use the .map extension function? 
```
channelState.channelConfig.asLiveData().map { config -> config.commands }
```

It is just suggar syntax, of course, but I think it is more aligned with the other Kotlin APIs (like Iterables)

**Review by leandroBorgesFerreira — COMMENTED** [review]
I think it is better to add the dependencies in the constructor of the class, instead of user ChatClient.logic, ChatClient.state, ChatClient.globalState... If we access the dependencies this way, all the classes with be super bond to ChatClient, which may make the code harder to refactor is we need to do it in the future (like we are in ChatDomain). 

Beside, constructor inject makes unit tests much easier to much, which is always good. 

What do you think?

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt (hunk 6 above)
Comment (MarinTolic) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt:73]: This is good advice.

Honestly, I know `LiveData` is usually operator poor so I forget that it has at least some operator capabilities.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt [hunk 7]
```diff
@@ -1,70 +1,99 @@
...
-    private val _activeThread = MutableLiveData<Message?>()
-    private val _members = MediatorLiveData<List<Member>>()
-    private val _channelState = MediatorLiveData<Channel>()
-    private val _anyOtherUsersOnline = MediatorLiveData<Boolean>()
-    private val _typingUsers = MediatorLiveData<List<User>>()
+    /**
+     * Holds information about the current channel and is actively updated.
+     */
+    public val channelState: ChannelState =
+        chatClient.asReferenced().watchChannel(cid, MessageListViewModel.DEFAULT_MESSAGES_LIMIT).asState(viewModelScope)
```
Comment (MarinTolic) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt:41]: Do we even need a concrete message limit when we watch channels in `ViewModels` that deal with headers, etc. and do not display the actual messages?

File: stream-chat-android-ui-common/src/test/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModelTest.kt [hunk 8]
```diff
@@ -2,33 +2,7 @@
 
 package com.getstream.sdk.chat.viewmodel
 
-import androidx.lifecycle.MutableLiveData
```
Comment (MarinTolic) [inline stream-chat-android-ui-common/src/test/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModelTest.kt:5]: Tests will be refactored in a separate PR so that the Core team is not blocked for further migration.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt (hunk 1 above)
Comment (MarinTolic) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt:58]: Matter resolved in a meeting.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt (hunk 2 above)
Comment (MarinTolic) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt:66]: Matter was resolved in a meeting. 
Decided on keeping the `LiveData`.

File: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/add/AddChannelViewModel.kt [hunk 9]
```diff
@@ -112,7 +110,7 @@ class AddChannelViewModel : ViewModel() {
             return
         }
         viewModelScope.launch(Dispatchers.IO) {
-            val currentUserId = chatDomain.user.value?.id ?: error("User must be set before create new channel!")
+            val currentUserId = chatClient.getCurrentUser()?.id ?: error("User must be set before create new channel!")
```
Comment (MarinTolic) [inline stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/add/AddChannelViewModel.kt:113]: Here we are not using `GlobalState`.
I've tested this offline and it appreas safe but stress test this as much as you can.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt (hunk 7 above)
Comment (adasiewiczr) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/BaseMessageListHeaderViewModel.kt:41]: We don't need any messages here. We should replace it with `0`

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt [hunk 10]
```diff
@@ -16,85 +17,127 @@ import io.getstream.chat.android.client.models.Command
...
-    private val chatDomain: ChatDomain = ChatDomain.instance(),
     private val chatClient: ChatClient = ChatClient.instance(),
+    globalState: GlobalState = chatClient.globalState,
 ) : ViewModel() {
+
+    /**
+     * Holds information about the current channel and is actively updated.
+     */
+    public val channelState: ChannelState =
+        chatClient.asReferenced().watchChannel(cid, MessageListViewModel.DEFAULT_MESSAGES_LIMIT).asState(viewModelScope)
```
Comment (adasiewiczr) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt:49]: I think we don't need any message here

File: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/info/group/users/GroupChatInfoAddUsersViewModel.kt [hunk 11]
```diff
@@ -4,24 +4,32 @@ import androidx.lifecycle.LiveData
...
 class GroupChatInfoAddUsersViewModel(
     cid: String,
-    chatDomain: ChatDomain = ChatDomain.instance(),
     chatClient: ChatClient = ChatClient.instance(),
 ) : ViewModel() {
 
+    /**
+     * Holds information about the current channel and is actively updated.
+     */
+    @OptIn(InternalStreamChatApi::class)
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/info/group/users/GroupChatInfoAddUsersViewModel.kt:29]: It's because `ChannelState` and `QueryChannelState` are annotated with `@InternalStreamChatAPI`, right? We should remove this annotation and those opt ins.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/actions/internal/ChannelActionsViewModel.kt [hunk 12]
```diff
@@ -5,51 +5,94 @@ import androidx.lifecycle.MediatorLiveData
...
     private val isGroup: Boolean,
-    chatDomain: ChatDomain = ChatDomain.instance(),
+    chatClient: ChatClient = ChatClient.instance(),
 ) : ViewModel() {
 
+    /**
+     * Holds information about the current channel and is actively updated.
+     */
+    val channelState: ChannelState =
+        chatClient.asReferenced().watchChannel(cid).asState(viewModelScope)
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/actions/internal/ChannelActionsViewModel.kt:34]: Do we need messages here? 🤔

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/typing/viewmodel/TypingIndicatorViewModel.kt [hunk 13]
```diff
@@ -1,30 +1,38 @@
...
+        chatClient.asReferenced().watchChannel(cid).asState(viewModelScope)
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/typing/viewmodel/TypingIndicatorViewModel.kt:29]: Do we need messages here? 🤔

**Review by adasiewiczr — APPROVED** [review]
Great job @MarinTolic. I've added some comments but we can resolve them in a separate PR as this one is quite big.

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/info/group/users/GroupChatInfoAddUsersViewModel.kt (hunk 11 above)
Comment (MarinTolic) [reply stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/info/group/users/GroupChatInfoAddUsersViewModel.kt:29]: Yup, `ChannelState` is marked as internal.

I can remove the annotations in a clean up PR if you want.
